File: Part 2 - Regression/multiple_linear_regression.R
# Multiple linear regression

# Importing the dataset
dataset = read.csv("50_Startups.csv")

# Encoding categorical data
dataset$State = factor(dataset$State, levels = c('New York','California','Florida'),
                       labels = c(1,2,3))


#install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)
training_set = subset(dataset,split ==TRUE)
test_set = subset(dataset,split ==FALSE)

# fitting multi reg on training set
regressor = lm(formula = Profit~ . , data = training_set)

# Predicting the results
y_pred = predict(regressor, newdata = test_set)

# Backward elimination
regressor = lm(formula = Profit~R.D.Spend + Administration + Marketing.Spend + State , 
               data = dataset)
summary(regressor)

regressor = lm(formula = Profit~R.D.Spend + Administration + Marketing.Spend , 
               data = dataset)
summary(regressor)


regressor = lm(formula = Profit~R.D.Spend + Marketing.Spend , 
               data = dataset)
summary(regressor)

regressor = lm(formula = Profit~R.D.Spend, 
               data = dataset)
summary(regressor)

File: Part 2 - Regression/multiple_linear_regression.py
# Multiple linear regression

# Importing the libraries
import numpy as np
import pandas as pd
import matplotlib as plt

# Importing the dataset
dataset = pd.read_csv("50_Startups.csv")
X = dataset.iloc[:, :-1].values
Y = dataset.iloc[:, 4].values

# Taking care of the missing values
from sklearn.preprocessing import Imputer, LabelEncoder, OneHotEncoder
"""
imputer = Imputer(missing_values='NaN', strategy="mean", axis=0)
imputer = imputer.fit(X[:, 1:3])
X[:, 1:3] = imputer.transform(X[:, 1:3])
"""

# Encoding categorical data
labelencoder_X = LabelEncoder()
X[:, 3] = labelencoder_X.fit_transform(X[:, 3])

onehotencoder = OneHotEncoder(categorical_features=[3])
X = onehotencoder.fit_transform(X).toarray()
X = X[:,1:]

# Spliting the dataset into trainingset and testset
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y,
                                                    test_size=0.2,
                                                    random_state=0)

"""
# feature scaling
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
"""

# Fitting the Regressor to the training set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train,Y_train)

# Predicting the result
y_pred = regressor.predict(X_test)

# Backward elimination
import statsmodels.formula.api as sm
X = np.append(arr =np.ones((50,1)).astype(int), values =X ,axis=1)
X_opt = X[:, [0,1,2,3,4,5]]
regressor_OLS = sm.OLS(endog=Y, exog= X_opt).fit()
regressor_OLS.summary()


X_opt = X[:, [0,1,3,4,5]]
regressor_OLS = sm.OLS(endog=Y, exog= X_opt).fit()
regressor_OLS.summary()

X_opt = X[:, [0,3,4,5]]
regressor_OLS = sm.OLS(endog=Y, exog= X_opt).fit()
regressor_OLS.summary()

X_opt = X[:, [0,3,5]]
regressor_OLS = sm.OLS(endog=Y, exog= X_opt).fit()
regressor_OLS.summary()

X_opt = X[:, [0,3]]
regressor_OLS = sm.OLS(endog=Y, exog= X_opt).fit()
regressor_OLS.summary()
